libafl/src/executors/inprocess/mod.rs

```diff
@@ -41,150 +41,162 @@ pub mod inner;
-    HB: BorrowMut<H>,
-    HT: ExecutorHooksTuple<S>,
-    OT: ObserversTuple<S>,
-    S: State,
-{
-    type State = S;
+impl<H, HT, OT> HasObservers for GenericInProcessExecutor<H, HT, OT> {
+    type Observers = OT;
+
+    #[inline]
+    fn observers(&self) -> RefIndexable<&Self::Observers, Self::Observers> {
+        RefIndexable::from(&self.observers)
+    }
+
+    #[inline]
+    fn observers_mut(&mut self) -> RefIndexable<&mut Self::Observers, Self::Observers> {
+        RefIndexable::from(&mut self.observers)
+    }
 }
 
-impl<H, HB, HT, OT, S> UsesObservers for GenericInProcessExecutor<H, HB, HT, OT, S>
-where
-    H: FnMut(&S::Input) -> ExitKind + ?Sized,
-    HB: BorrowMut<H>,
-    HT: ExecutorHooksTuple<S>,
-    OT: ObserversTuple<S>,
-    S: State,
-{
-    type Observers = OT;
+#[derive(Debug)]
+pub struct RecoveryHook;
+
+impl<E, EM, I, S, Z> ExecutorHook<E, EM, I, S, Z> for RecoveryHook {
+    fn pre_exec(
+        &mut self,
+        executor: &mut E,
+        fuzzer: &mut Z,
+        state: &mut S,
+        mgr: &mut EM,
+        input: &S,
+    ) {
+        todo!()
+    }
+
+    fn post_exec(
+        &mut self,
+        executor: &mut E,
+        fuzzer: &mut Z,
+        state: &mut S,
+        mgr: &mut EM,
+        input: &S,
+    ) {
+        todo!()
+    }
 }
 
-impl<EM, H, HB, HT, OT, S, Z> Executor<EM, Z> for GenericInProcessExecutor<H, HB, HT, OT, S>
+/// This function marks the boundary between the fuzzer and the target for in-process executors
+///
+/// # Safety
+/// This function sets a bunch of raw pointers in global variables, reused in other parts of
+/// the code.
+#[inline]
+pub fn enter_target<EM, I, S, Z>(
+    fuzzer: &mut Z,
+    state: &mut S,
+    mgr: &mut EM,
+    input: &S,
+    executor_ptr: *const c_void,
+) {
+    unsafe {
+        let data = addr_of_mut!(GLOBAL_STATE);
+        write_volatile(
+            addr_of_mut!((*data).current_input_ptr),
+            ptr::from_ref(input) as *const c_void,
+        );
+        write_volatile(addr_of_mut!((*data).executor_ptr), executor_ptr);
+        // Direct raw pointers access /aliasing is pretty undefined behavior.
+        // Since the state and event may have moved in memory, refresh them right before the signal may happen
+        write_volatile(
+            addr_of_mut!((*data).state_ptr),
+            ptr::from_mut(state) as *mut c_void,
+        );
+        write_volatile(
+            addr_of_mut!((*data).event_mgr_ptr),
+            ptr::from_mut(mgr) as *mut c_void,
+        );
+        write_volatile(
+            addr_of_mut!((*data).fuzzer_ptr),
+            ptr::from_mut(fuzzer) as *mut c_void,
+        );
+        compiler_fence(Ordering::SeqCst);
+    }
+}
+
+/// This function marks the boundary between the fuzzer and the target
+#[inline]
+pub fn leave_target() {
+    unsafe {
+        let data = addr_of_mut!(GLOBAL_STATE);
+
+        write_volatile(addr_of_mut!((*data).current_input_ptr), ptr::null());
```

Comment:
Clearly also unsafe, right?

Reply:
Yup, but we'll be changing this out soon anyways.